### PR TITLE
explorer: overhaul rule generator caching and matching

### LINKED ISSUE #1159: rule generator: `instruction` scope support/bug
Need to verify implementation:

62577A7CC79DF01C730F66797D00228E:0x1018DB6 should match

### LINKED ISSUE #1246: capa explorer: AttributeError: "CapaExplorerForm" object has no attribute "rules_cache" (load_capa_function_results)
I think this is a result of removing the `None` initialization in #1235 

error occurs when you first open the plugin, rule generator tab, and select `Analyze`

### PR DESCRIPTION
**Pull request by mike-hunhoff**
- [x] fix #1246 
- [x] fix #1159 
- [x] fix rule generator `MatchedRule` matching
- [x] enable file scope matching
- [x] refactor form code to remove circular reference
- [x] fix scope matching
- [x] add "instruction" scope to UI nesting options

Screen Shot 2023-01-03 at 10 18 26 AM

### CODE REVIEW

File: capa/ida/plugin/extractor.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Copyright (C) 2020 Mandiant, Inc. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
```
Comment (mike-hunhoff) [inline capa/ida/plugin/extractor.py:3]: move to new file - avoid circular reference

File: capa/ida/plugin/error.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Copyright (C) 2020 Mandiant, Inc. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
```
Comment (mike-hunhoff) [inline capa/ida/plugin/error.py:3]: move to new file - avoid circular reference

**Comment by mr-tz** [comment]
Please see https://github.com/mandiant/capa/pull/1257 for an explorer improvement that depends on the fixes here.

File: capa/ida/plugin/cache.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+
+class CapaRuleGenFeatureCache:
+    def __init__(self, fh_list: List[FunctionHandle], extractor: CapaExplorerFeatureExtractor):
+        self.global_features: FeatureSet = collections.defaultdict(set)
+
+        self.file_node: CapaRuleGenFeatureCacheNode = CapaRuleGenFeatureCacheNode(None, None)
+        self.func_nodes: Dict[Address, CapaRuleGenFeatureCacheNode] = {}
+        self.bb_nodes: Dict[Address, CapaRuleGenFeatureCacheNode] = {}
+        self.insn_nodes: Dict[Address, CapaRuleGenFeatureCacheNode] = {}
```
Comment (mike-hunhoff) [inline capa/ida/plugin/cache.py:65]: Approach: the rule generator caches features per function at each scope using tree structure. This allows us to quickly and correctly check rule matches at specific scopes without needing to re-emit the features.

File: capa/ida/plugin/extractor.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+class CapaExplorerProgressIndicator(QtCore.QObject):
+    """implement progress signal, used during feature extraction"""
+
+    progress = QtCore.pyqtSignal(str)
+
+    def __init__(self):
+        """initialize signal object"""
+        super().__init__()
+
```
Comment (williballenthin) [inline capa/ida/plugin/extractor.py:25]:
```suggestion
```

**Review by mr-tz — COMMENTED** [review]
just minor comments, thanks!

File: capa/ida/plugin/extractor.py (hunk 4 above)
Comment (mike-hunhoff) [reply capa/ida/plugin/extractor.py:25]: good catch!

**Pull request merged**